### ISSUE TITLE
Can we have a release version along with GitHub Actions?

### ISSUE THREAD

**Issue by withoutK** [UniMath/SymmetryBook#119]

**Comment by UlrikBuchholtz** [comment]
Yes, we're already building the book on every push, we just need to use jekyll to put it on gh-pages. I'll assign @pierrecagne for this.

**Comment by UlrikBuchholtz** [comment]
We also need to typeset the commit hash, so it's easy to refer to. I'll look in to this.

**Comment by DanGrayson** [comment]
We could even include that commit number in a recommended bibtex item,
included in the book.

**Comment by UlrikBuchholtz** [comment]
In the latest commit c22710a I sort of crudely include the commit information from a file, `version.tex`, that is generated by make. It's a simple solution, but it won't update until you remove that file, then make will remake.

**Comment by DanGrayson** [comment]
Maybe the dependencies I've just added in commit f374c0d844bca057ff2ab90698cd72b50917881f will get version.tex remade when necessary.

**Comment by withoutK** [comment]
thx :)

**Comment by DanGrayson** [comment]
Maybe the github action should simply run "make":

https://github.com/UniMath/SymmetryBook/runs/3374257414?check_suite_focus=true#step:5:26

**Comment by favonia** [comment]
> Maybe the dependencies I've just added in commit [f374c0d](https://github.com/UniMath/SymmetryBook/commit/f374c0d844bca057ff2ab90698cd72b50917881f) will get version.tex remade when necessary.

I would recommend against dependency on `FETCH_HEAD`. It means "the last commit fetched by git", with can be different from `HEAD`. Since `--shell-escape` is already enabled, an alternative approach is to call `git` directly from within LaTaX. This would avoid the problem of keeping the intermediate file `version.tex` up to date.

Another thing is that `"...%ad..." --date=short` can be abbreviated to `"...%as..."`.

**Comment by DanGrayson** [comment]
That sounds like a good idea.

**Comment by UlrikBuchholtz** [comment]
Thanks @favonia ! I've updated the dependency to be on `.git/refs/heads/master` – should be alright for our purposes.

I also set things up so TikZ externalization is disabled in the GH action, reducing the build time by more than half.

And (@pierrecagne), I saw an easy way to use GH actions to deploy the result to gh-pages, so now that's done too (and only on commits to master).

I like the approach with `version.tex` because then we can compile most of the time with a standard `pdflatex` (we only need `--shell-escape` for TikZ externalization).

**Comment by pierrecagne** [comment]
Ok, seems you beat me to it. I saw yesterday a GH action very similar to the one you use and wanted to take care of that tomorrow when time permits. 

One thing left to do before closing the issue is to address Dan's comment: why use dante-ev/latex-action@latest for compiling when we can just run make? Any advantages in using this github action?

**Comment by pierrecagne** [comment]
Also I like the idea to include in the book (on the second page for example) a "How to cite", providing a BibTeX excerpt with the commit number inside, so that people can simply copy-paste to their bib file.

**Comment by UlrikBuchholtz** [comment]
I'm not sure if TeX is even present without using that action, which specifies a Docker image with the latest TeXlive. But we can try.

I agree about putting some BibTeX on a copyright page (p. 2), which we should have anyway.

**Comment by pierrecagne** [comment]
Yes of course, sorry. What I meant is the following: could we just load the action, so we have access to the Docker with TeXLive, but then use our make command in actually compiling? I'm only worried that if one changes the make command in the future and forget to mirror the changes in the call of the github action, it will end up with a version online not matching the local build (or even not compiling). There seems to be room for that in dante-ev's action: from the README 
```
- name: release.sh
        uses: dante-ev/latex-action@latest
        with:
          entrypoint: ./release.sh
```

**Comment by UlrikBuchholtz** [comment]
I'd rather make sure that `latexmk` always knows how to make the book (we use `latexmkrc` for this), so I'm not too worried about that. Also, I implemented the disabling of TikZ externalize by adding something to version.tex after `make` makes it, so that approach wouldn't work if we just ran `make`. (We could use a flag to `make` instead.)

So I vote to close this issue now; I'll make a new issue for the copyright page.

**Comment by DanGrayson** [comment]
But I suppose `latexmk` will never know when it ought to remake `version.tex`.  Or can it?

**Comment by pierrecagne** [comment]
> But I suppose `latexmk` will never know when it ought to remake `version.tex`. Or can it?

I think that's fine: every time somebody pushes new commits, a virtual machine is created somewhere in Github's cloud system and run all the commands from scratch, one of the first being `make version.tex`. (The actual process is probably much more complex , but this is observationally equivalent from our point of view.)

**Comment by favonia** [comment]
> But I suppose `latexmk` will never know when it ought to remake `version.tex`. Or can it?

This is not entirely impossible but would involve ugly hacks that I believe no one wants to touch. On the other hand, I personally am fine with my local copies showing wrong commit hashes. The public versions will always be correct.

**Comment by UlrikBuchholtz** [comment]
Yes, I see that as the key point, too: The version on gh-pages should be correct. You can have `\newcommand{\OPTversion}{working copy}` or whatever you like in your own `version.tex`.

BTW, the deployment action makes fresh commits to gh-pages and doesn't preserve any history. In particular, we lost the old html files, so I've now moved them to the main branch and made sure they're included in the build process. So now the links in the README file work again. I also added a link to the book there, so people can find it.

**Comment by Iey4iej3** [comment]
> Another thing is that `"...%ad..." --date=short` can be abbreviated to `"...%as..."`.

`%as` is not supported on `2.20.1` (Debian Buster). It might be better that the project could be compiled on a not-so-up-to-date platform.

**Comment by UlrikBuchholtz** [comment]
This is now fixed in commit 7e714ef – I just tested that the build process works on a machine with Debian 8.11 and TeX Live 2015.
